### PR TITLE
394 enhance engines to provide unified workflow

### LINKED ISSUE #394: enhance engines for future extension
**Is your feature request related to a problem? Please describe.**
Our existing ignite engine under /engines is developed as the first component of this project.
As we already have many new features, we need to update the engines for future extension.
Users also asked about the multi-gpu engine and the example.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #394 .

### Description
This PR enhanced our `engines` module with `trainers` and `evaluators`.

They can provide **unified** and **easier** training and evaluation workflows, it will be useful to set advanced strategies, like: JSON config, AutoML and Federated Learning, etc.

For next steps, I will also do these items in other PRs:
1. Update and enhance our handlers, add more useful handlers.
2. Develop workflow integration test.

### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: monai/engines/workflow.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+import torch
+from ignite.engine import Engine, State, Events
+from .utils import default_prepare_batch
+
+
+class Workflow(ABC):
```
Comment (ericspod) [inline monai/engines/workflow.py:18]: I'd suggest this class should inherit `ignite.engine.Engine` instead of having an internally created instance of it. This would simplify code so that `self.engine` wouldn't be needed everywhere, and any existing code expecting `Engine` instances will function just by passing in a `Workflow` instance without having to type check.

File: monai/engines/workflow.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        key_metric (ignite.metric): compute metric when every iteration completed, and save average
+            value to engine.state.metrics when epoch completed. also use key_metric to select and
+            save checkpoint into files.
+        additional_metrics (list): more ignite metrics that also attach to Ignite Engine.
+        handlers (list): every handler is a set of Ignite Event-Handlers, like:
+            CheckpointHandler, StatsHandler, TimerHandler, etc.
+
+    """
+
+    def __init__(
```
Comment (ericspod) [inline monai/engines/workflow.py:40]: It would be nice to have an optional callable parameter which is used as the step function for the `Engine`, if not given `self._iteration` is used as the default.

File: monai/engines/workflow.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        amp,
+        data_loader,
+        prepare_batch=default_prepare_batch,
+        key_metric=None,
+        additional_metrics=None,
+        handlers=None,
+    ):
+        # FIXME:
+        if amp:
+            print("Will add AMP support when PyTorch v1.6 released.")
```
Comment (ericspod) [inline monai/engines/workflow.py:53]: This should be a log message.

File: monai/engines/workflow.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        assert isinstance(device, torch.device), "must provide PyTorch device information."
```
Comment (ericspod) [inline monai/engines/workflow.py:54]: Instead of assert, a conditional should be used to validate inputs which raise `ValueError` if the condition isn't met.

File: monai/engines/workflow.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            def post_epoch_process(engine):
+                if engine.state.key_metric_name is not None:
+                    current_val_metric = engine.state.metrics[engine.state.key_metric_name]
+                    if current_val_metric > engine.state.best_metric:
+                        print(f"Got new best metric of {engine.state.key_metric_name}: {current_val_metric}")
+                        engine.state.best_metric = current_val_metric
+                        engine.state.best_metric_epoch = engine.state.epoch
+
+        if handlers is not None and len(handlers) > 0:
+            assert isinstance(handlers, (list, tuple)), "handlers must be a chain."
```
Comment (ericspod) [inline monai/engines/workflow.py:100]: If `handlers` is not iterable the for loop below will raise an exception so this check isn't strictly necessary and doesn't permit other types besides list or tuple.

**Comment by ericspod** [comment]
There is some discussion with Ignite on high level API types that has overlap with this PR: https://github.com/pytorch/ignite/issues/912

**Comment by Nic-Ma** [comment]
> There is some discussion with Ignite on high level API types that has overlap with this PR: [pytorch/ignite#912](https://github.com/pytorch/ignite/issues/912)

Hi @ericspod ,

Thanks very much for your review and sharing!
I will totally adjust the workflows tomorrow based on your previous investigation and comments.
Maybe we can use this PR as the baseline and then enhance & add more workflows later.
I think the unified training workflow will be very useful, especially for JSON/YAML config, AutoML and Federated Learning, etc.
Thanks.

**Comment by vfdev-5** [comment]
Thanks for mentioning @wyli ! Yes, definitely, classes like `SupervisedTrainer` make a lot of sense and it is something we aim to provide in Ignite too while keeping the flexibility of Engine.

PS: I also cc @sdesrozis as @justusschock is not working anymore on Ignite.

File: monai/engines/workflow.py (hunk 5 above)
Comment (Nic-Ma) [reply monai/engines/workflow.py:100]: removed the assert in the latest patch.
Thanks.

File: monai/engines/workflow.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/engines/workflow.py:18]: updated to inherit Engine in the latest patch.
Thanks.

File: monai/engines/workflow.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/engines/workflow.py:53]: updated in the latest patch.
Thanks.

File: monai/engines/workflow.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/engines/workflow.py:40]: added step_function in the latest patch.
Thanks.

File: monai/engines/workflow.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/engines/workflow.py:54]: updated in the latest patch.
Thanks.

File: monai/engines/utils.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+
+
+class CommonKeys:
+    IMAGE = "image"
+    LABEL = "label"
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/engines/utils.py:17]: updated in the latest patch.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod and @wyli ,

Thanks very much for your review and feedback!
I updated the PR according to your comments.
Could you please help review it again?
Thanks.

File: monai/handlers/validation_handler.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ignite.engine import Events
+
+
+class ValidationHander:
```
Comment (sdesrozis) [inline monai/handlers/validation_handler.py:15]: Typo error ? `ValidationHandler`

File: tests/test_handler_validation.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        # set up engine
+        def _train_func(engine, batch):
+            pass
+
+        engine = Engine(_train_func)
+
+        # set up testing handler
+        val_data_loader = torch.utils.data.DataLoader(Dataset(data))
+        evaluator = TestEvaluator(torch.device("cpu:0"), val_data_loader)
+        saver = ValidationHander(evaluator, interval=2)
```
Comment (sdesrozis) [inline tests/test_handler_validation.py:38]: Typo again about ˋValidationHandler` ?

**Comment by sdesrozis** [comment]
Hi ! I reviewed (quite late) this PR because closely related to ignite API. I really appreciate the smart usage of ignite done here. it makes me think that ignite is exactly in the right position for projects like MONAI : providing building blocks to compose complex applications.

My feedback about the PR is maybe we need in ignite a way to define smart functions in `Engine`, a bit more than a `Callable` to help about `amp` for instance. Moreover, I think that the work we are currently doing about agnostic distributed computing (for tpu, cuda, etc.) should be helpful for MONAI. We will support on that way if you need 👍🏻

File: monai/handlers/validation_handler.py (hunk 7 above)
Comment (Nic-Ma) [reply monai/handlers/validation_handler.py]: updated in the latest patch.
Thanks.

File: tests/test_handler_validation.py (hunk 8 above)
Comment (Nic-Ma) [reply tests/test_handler_validation.py]: updated in the latest patch.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @sdesrozis ,

Thanks very much for your review and feedback from ignite side, glad to co-work with you guys.
I updated the PR according to your comments.
And we will add AMP support when PyTorch v1.6 released(it's just reserved for API now).
We may leverage the AMP feature in ignite directly.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod and @wyli ,

I updated this PR with all the latest comments, could you please help review it again?
Thanks.

**Comment by Nic-Ma** [comment]
Thanks for @wyli 's review, I updated according to his offline comments:
1. change `_run()` to `run()`.
2. delete the `_iteration()` abstract function in Trainer and Evaluator.
3. add doc-string for the `global_epoch` parameter of Evaluator.
4. change `RegularInferer` to `SimpleInferer`.
5. fix several small typos in doc-strings and documentation.
6. refine all the doc-strings in this PR.

I also changed all the `assert` to `raise Exception` according to @ericspod 's comments.

@ericspod , @wyli , Could you please help review it again?
Thanks in advance.

File: monai/engines/evaluator.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        iteration_update=None,
+        key_val_metric=None,
+        additional_metrics=None,
+        val_handlers=None,
+    ):
+        super().__init__(
+            device,
+            1,
+            False,
+            val_data_loader,
```
Comment (wyli) [inline monai/engines/evaluator.py:52]: could use `device=device, max_epochs=1, amp=False, ... ` for code readability (same for the trainers)

**Comment by Nic-Ma** [comment]
Thanks @wyli for your review!

Hi @ericspod ,

Do you have any other comments or suggestions?
Thanks.

**Comment by ericspod** [comment]
It looks good in general. The approach I had taken with a higher level `Engine` type was to have a base type inheriting `Engine` which served the role of an inferer as well as being subtyped as a `Trainer` and `Evaluator` classes to suite those roles. The idea was that this provided a simple inference capability regardless of which subtype was used, so one could have inference routines that could work with either. 

Another idea was to have a method implementing the network forward pass (pass arguments to the network, packaging results into a tuple) which both subtypes would use, this would reduce code and allow adapting the network to data parallel operations without creating a wrapper type. Similarly there was a method for wrapping the forward pass for the loss function. The idea with this was to allow a subtype to modify how the forward passes worked without having to change any of the other code. 

I'm fine with merging this now though, these are ideas for evolution into the future perhaps.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Thanks for your review and for sharing cool ideas!
I think your 2 proposals are very interesting, they are mainly focusing on the optimization for the operations of `inferer` and `network`or `loss`, which are not enhanced in this PR. This PR is mainly to enhance the operations for `iteration`, `handlers`, and `metrics` in the ignite engines. 
So maybe we can work together later to **combine** your proposals with current Trainer and Evaluator. I will merge this PR first and develop other higher level modules in later PRs.

Thanks.
